File: backend/src/main/java/com/petyilgi/pet/domain/Pet.java
package com.petyilgi.pet.domain;

import com.petyilgi.auth.domain.User;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.SQLRestriction;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "pets")
@SQLRestriction("is_deleted = false")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Pet {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 50)
    private String name;

    @Column(nullable = false, length = 30)
    private String species;

    @Column(name = "birth_date")
    private LocalDate birthDate;

    @Column(length = 80)
    private String breed;

    @Column(name = "adoption_date")
    private LocalDate adoptionDate;

    @Enumerated(EnumType.STRING)
    private Gender gender;

    @Column(precision = 5, scale = 2)
    private BigDecimal weight;

    @Column(name = "animal_registration_number", length = 20)
    private String animalRegistrationNumber;

    private Boolean neutered;

    @Column(columnDefinition = "TEXT")
    private String diseases;

    @Column(name = "special_notes", columnDefinition = "TEXT")
    private String specialNotes;

    @Column(name = "guardian_nickname", length = 30)
    private String guardianNickname;

    @Column(name = "special_status", length = 30)
    private String specialStatus;

    @Column(columnDefinition = "TEXT")
    private String personality;

    @Column(name = "primary_hospital_name", length = 100)
    private String primaryHospitalName;

    @Column(name = "accent_color", nullable = false, length = 7)
    private String accentColor;

    @Column(name = "bg_light", nullable = false, length = 7)
    private String bgLight;

    @Column(name = "is_deleted", nullable = false)
    private boolean deleted = false;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    public static Pet create(User user, String name, String species, LocalDate birthDate,
                             String accentColor, String bgLight) {
        Pet pet = new Pet();
        pet.user        = user;
        pet.name        = name;
        pet.species     = species;
        pet.birthDate   = birthDate;
        pet.accentColor = accentColor;
        pet.bgLight     = bgLight;
        pet.createdAt   = LocalDateTime.now();
        pet.updatedAt   = LocalDateTime.now();
        return pet;
    }

    public void updateName(String name) {
        this.name      = name;
        this.updatedAt = LocalDateTime.now();
    }

    public void update(String name, String species, LocalDate birthDate,
                       Gender gender, BigDecimal weight, String animalRegistrationNumber,
                       Boolean neutered, String diseases, String specialNotes,
                       String breed, LocalDate adoptionDate, String guardianNickname,
                       String specialStatus, String personality, String primaryHospitalName,
                       String accentColor, String bgLight) {
        this.name                       = name;
        this.species                    = species;
        this.birthDate                  = birthDate;
        this.gender                     = gender;
        this.weight                     = weight;
        this.animalRegistrationNumber   = animalRegistrationNumber;
        this.neutered                   = neutered;
        this.diseases                   = diseases;
        this.specialNotes               = specialNotes;
        this.breed                      = breed;
        this.adoptionDate               = adoptionDate;
        this.guardianNickname           = guardianNickname;
        this.specialStatus              = specialStatus;
        this.personality                = personality;
        this.primaryHospitalName        = primaryHospitalName;
        this.accentColor                = accentColor;
        this.bgLight                    = bgLight;
        this.updatedAt                  = LocalDateTime.now();
    }

    public void softDelete() {
        this.deleted   = true;
        this.updatedAt = LocalDateTime.now();
    }

    public boolean isOwnedBy(Long userId) {
        return user.getId().equals(userId);
    }
}

File: frontend/lib/models/pet.dart
class Pet {
  final String id;
  final String name;
  final String species;
  final String? birthDate;
  final String? breed;
  final String? adoptionDate;
  final String accentColor;
  final String bgLight;
  final String? gender;
  final double? weight;
  final String? animalRegistrationNumber;
  final bool? neutered;
  final String? specialNotes;
  final String? diseases;
  final String? guardianNickname;
  final String? specialStatus;
  final String? personality;
  final String? primaryHospitalName;
  final String? profileImageUrl;

  const Pet({
    required this.id,
    required this.name,
    required this.species,
    required this.birthDate,
    this.breed,
    this.adoptionDate,
    required this.accentColor,
    required this.bgLight,
    this.gender,
    this.weight,
    this.animalRegistrationNumber,
    this.neutered,
    this.specialNotes,
    this.diseases,
    this.guardianNickname,
    this.specialStatus,
    this.personality,
    this.primaryHospitalName,
    this.profileImageUrl,
  });

  factory Pet.fromJson(Map<String, dynamic> j) => Pet(
    // Backend returns Long id — convert to String
    id: j['id'].toString(),
    name: j['name'] as String,
    species: j['species'] as String,
    birthDate: j['birthDate'] as String?,
    breed: j['breed'] as String?,
    adoptionDate: j['adoptionDate'] as String?,
    accentColor: j['accentColor'] as String? ?? '#FF8A65',
    bgLight: j['bgLight'] as String? ?? '#FFF3E0',
    // gender enum: male | female (lowercase) — may be null
    gender: j['gender'] as String?,
    weight: j['weight'] != null
        ? double.tryParse(j['weight'].toString())
        : null,
    animalRegistrationNumber: j['animalRegistrationNumber'] as String?,
    neutered: j['neutered'] as bool?,
    specialNotes: j['specialNotes'] as String?,
    diseases: j['diseases'] as String?,
    guardianNickname: j['guardianNickname'] as String?,
    specialStatus: j['specialStatus'] as String?,
    personality: j['personality'] as String?,
    primaryHospitalName: j['primaryHospitalName'] as String?,
    profileImageUrl: j['profileImageUrl'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'name': name,
    'species': species,
    if (birthDate != null) 'birthDate': birthDate,
    if (breed != null) 'breed': breed,
    if (adoptionDate != null) 'adoptionDate': adoptionDate,
    'accentColor': accentColor,
    'bgLight': bgLight,
    if (gender != null) 'gender': gender,
    if (weight != null) 'weight': weight,
    if (animalRegistrationNumber != null)
      'animalRegistrationNumber': animalRegistrationNumber,
    if (neutered != null) 'neutered': neutered,
    if (specialNotes != null) 'specialNotes': specialNotes,
    if (diseases != null) 'diseases': diseases,
    if (guardianNickname != null) 'guardianNickname': guardianNickname,
    if (specialStatus != null) 'specialStatus': specialStatus,
    if (personality != null) 'personality': personality,
    if (primaryHospitalName != null) 'primaryHospitalName': primaryHospitalName,
    if (profileImageUrl != null) 'profileImageUrl': profileImageUrl,
  };

  Pet copyWith({
    String? name,
    String? species,
    String? birthDate,
    String? breed,
    String? adoptionDate,
    String? accentColor,
    String? bgLight,
    String? gender,
    double? weight,
    String? animalRegistrationNumber,
    bool? neutered,
    String? specialNotes,
    String? diseases,
    String? guardianNickname,
    String? specialStatus,
    String? personality,
    String? primaryHospitalName,
    String? profileImageUrl,
  }) => Pet(
    id: id,
    name: name ?? this.name,
    species: species ?? this.species,
    birthDate: birthDate ?? this.birthDate,
    breed: breed ?? this.breed,
    adoptionDate: adoptionDate ?? this.adoptionDate,
    accentColor: accentColor ?? this.accentColor,
    bgLight: bgLight ?? this.bgLight,
    gender: gender ?? this.gender,
    weight: weight ?? this.weight,
    animalRegistrationNumber:
        animalRegistrationNumber ?? this.animalRegistrationNumber,
    neutered: neutered ?? this.neutered,
    specialNotes: specialNotes ?? this.specialNotes,
    diseases: diseases ?? this.diseases,
    guardianNickname: guardianNickname ?? this.guardianNickname,
    specialStatus: specialStatus ?? this.specialStatus,
    personality: personality ?? this.personality,
    primaryHospitalName: primaryHospitalName ?? this.primaryHospitalName,
    profileImageUrl: profileImageUrl ?? this.profileImageUrl,
  );
}
